lib/ProductOpener/CRM.pm

```diff
@@ -0,0 +1,572 @@
+			{
+				x_off_org_id => $org_ref->{org_id},
+				category_id => [[4, $Odoo_tags{Producer}]],
+				x_off_main_contact => $user_ref->{crm_user_id}
+			}
+		]
+	);
+	return $req;
+}
+
+=head2 find_company ($name)
+
+Finds the oldest company following this strategy:
+
+1. if the company has the corresponding off_org_id, use this one
+2. if the contact_id is defined and the company he belongs to has no off_org_id, use this one 
+3. if there is a company with the exact same name, use this one
+
+=head3 Arguments
+
+=head4 $user_ref
+
+=head4 $org_ref
+
+=head3 Return values
+
+the company if found, undef otherwise
+
+=cut
+
+sub find_company($org_ref, $contact_id = undef) {
+	# 1. & 3. merged in one query
+	my $companies = Odoo(
+		'res.partner',
+		'search_read',
+		[
+			[
+				'&', ['is_company', '=', 1],
+				'|', ['x_off_org_id', '=', $org_ref->{org_id}],
+				'&',
+				['name', '=', $org_ref->{name}],
+				['x_off_org_id', 'like', '0']
+			]
+		],
+		{fields => ['name', 'id', 'x_off_org_id', 'is_company'], order => 'create_date ASC'}
+	);
+	my $company = $companies->[0];
+	# 1.
+	if (    defined $contact_id
+		and defined $company->{id}
+		and $company->{x_off_org_id} ne $org_ref->{org_id})
+	{
+		# find the company of the contact
+		my $req = Odoo('res.partner', 'read', [$contact_id], {fields => ['name', 'id', 'parent_id']});
+		# check if the company has no off_org_id
+		my $contact = $req->[0];
+
+		if (defined $contact && $contact->{parent_id} ne '0') {
+			my $req
+				= Odoo('res.partner', 'read', [$company->{parent_id}->[0]], {fields => ['name', 'id', 'x_off_org_id']});
+			my $contact_company = $req->[0];
+			if (defined $contact_company and $company->{x_off_org_id} eq '0') {
+				# 2.
+				return $contact_company->{id};
+			}
+		}
+		# 3
+	}
+	return $company->{id};
+}
+
+=head2 create_company ($org_ref)
+
+Creates a new company in Odoo from an org
+
+=head3 Arguments
+
+=head4 $org_ref
+
+=head3 Return values
+
+the id of the created company
+
+=cut
+
+sub create_company ($org_ref) {
+	my $main_contact_user_ref = retrieve_user($org_ref->{main_contact});
+	my $company = {
+		name => $org_ref->{name},
+		phone => $org_ref->{phone},
+		email => $org_ref->{email},
+		website => $org_ref->{website},
+		category_id => [$Odoo_tags{Producer}],    # "Producer" category id in Odoo
+		is_company => 1,
+		x_off_org_id => $org_ref->{org_id},
+		x_off_main_contact => $main_contact_user_ref->{crm_user_id},
+	};
+	my $company_id = Odoo('res.partner', 'create', [{%$company}]);
+	$log->debug("create_company", {company_id => $company_id, company => $company}) if $log->is_debug();
+	return $company_id;
+}
+
+=head2 add_contact_to_company ($org_ref)
+
+Add a contact to a company in Odoo
+
+=head3 Arguments
+
+=head4 $contact_id
+
+=head4 $company_id
+
+=head3 Return values
+
+1 if success, undef otherwise
+
+=cut
+
+sub add_contact_to_company($contact_id, $company_id) {
+	# set contact in children of company
+	# 4 means add an existing record
```

Comment:
Ah, here is my answer for the 4. :) We could set some constants for actions like this.